src/data/markdown/translated-guides/en/07 Misc/05 k6 Extensions.md

```diff
@@ -1,4 +1,400 @@
+For example, if a system you need support for can be tested with a generic protocol
+like MQTT, prefer using [xk6-mqtt](https://github.com/pmalhaire/xk6-mqtt)
+instead of creating an extension that uses some custom protocol.
+
+Next, you should decide the type of extension you need. A JavaScript extension is a
+good fit if you want to extend the JS functionality of your script, or add support
+for a new network protocol to test with. An Output extension would be more suitable
+if you need to process the metrics emitted by k6 in some way, submit them to a
+specific storage backend that was previously unsupported, etc. The k6 APIs you'll
+need to use and things to consider while developing will be different in each case.
+
+
+### Writing a new JavaScript extension
+
+A good starting point for using xk6 and writing a JS extension is the [xk6
+introductory article](https://k6.io/blog/extending-k6-with-xk6), but we'll cover
+some of the details here.
+
+JavaScript extensions consist of a main module struct that exposes methods that
+can be called from a k6 test script. For example:
+
+<!-- TODO: A better trivial example? -->
+
+<CodeGroup labels={["compare.go"]} lineNumbers={[false]}>
+
+```go
+package compare
+
+type Compare struct{}
+
+func (*Compare) IsGreater(int a, b) bool {
+	return a > b
+}
+```
+
+</CodeGroup>
+
+In order to use this from k6 test scripts we need to register the module
+by adding the following:
+
+<CodeGroup labels={["compare.go"]} lineNumbers={[false]}>
+
+```go
+import "go.k6.io/k6/js/modules"
+
+func init() {
+	modules.Register("k6/x/compare", new(Compare))
+}
+```
+
+</CodeGroup>
+
+Note that all k6 extensions should have the `k6/x/` prefix and the short name
+must be unique among all extensions built in the same k6 binary.
+
+We can then build a k6 binary with this extension by running
+`xk6 build --with xk6-compare=.`. In this case `xk6-compare` is the
+Go module name passed to `go mod init`, but in a real-world scenario
+this would be a URL.
+
+Finally we can use the extension in a test script:
+
+<CodeGroup labels={["test.js"]} lineNumbers={[true]}>
+
+```go
+import compare from 'k6/x/compare';
+
+export default function () {
+  console.log(compare.isGreater(2, 1));
+}
+```
+
+</CodeGroup>
+
+And run the test with `./k6 run test.js`, which should output `INFO[0000] true`.
+
+
+#### Additional features
+
+The k6 Go-JS bridge has a few features we should highlight:
+
+- Go method names will be converted from Pascal case to Camel case when
+  accessed in JS, as in the example above: `IsGreater` becomes `isGreater`.
+
+<!-- FIXME: Markdown magic to escape backticks in Go struct tags? <pre> doesn't work... -->
+- Similarly, Go field names will be converted from Pascal case to Snake case.
+  For example, the struct field `SomeField string` is accessible in JS
+  as the `some_field` object property. This behavior is configurable with the
+  `js` struct tag, so this can be changed with `SomeField string ``js:"someField"```
```

Comment:
the backtick escape is not working on staging

Reply:
Yeah :disappointed: I changed it to `<code>SomeField string &grave;js:"someField"&grave;</code>` in ecc287f, hopefully it renders fine :man_shrugging:

Reply:
![Screenshot from 2021-10-06 18-55-26](https://user-images.githubusercontent.com/2103732/136249237-85bc1bbb-92bf-46ea-92d7-5c9b345a524a.png)
I see this